R/bundlePackagePackrat.R

```diff
@@ -0,0 +1,200 @@
+  repos <- setNames(
+    sapply(repos, "[[", 2),
+    sapply(repos, "[[", 1)
+  )
+
+  # get packages records defined in the lockfile
+  records <- utils::tail(df, -1)
+  if (nrow(records) == 0) {
+    return(data.frame())
+  }
+
+  rownames(records) <- NULL
+  records <- records[manifestPackageColumns(records)]
+  records[c("Source", "Repository")] <- standardizeRecords(records, repos)
+
+  records$description <- lapply(records$Package, function(pkg) {
+    readLines(system.file("DESCRIPTION", package = pkg))
+  })
+
+  records
+}
+
+standardizeRecords <- function(records, repos) {
+  availablePackages <- availablePackages(repos)
+  repos <- standardizeRepos(repos)
+
+  rows <- lapply(seq_len(nrow(records)), function(i) {
+    standardizePackratPackage(records[i, ], availablePackages, repos = repos)
+  })
+  rows <- lapply(rows, as.data.frame, stringsAsFactors = FALSE)
+  rbind_fill(rows, c("Source", "Repository"))
+}
+
+standardizeRepos <- function(repos) {
+  # Ensure that each repository has a unique name
+  names(repos) <- ifelse(
+    names2(repos) == "",
+    paste0("repo_", seq_along(repos)),
+    names2(repos)
+  )
+
+  # And strip trailing /
+  repos <- gsub("/$", "", repos)
+
+  repos
+}
+
+standardizePackratPackage <- function(record, availablePackages, repos = character()) {
+  pkg <- record$Package
+  source <- record$Source
+
+  # source types are defined by packrat:
+  # https://github.com/rstudio/packrat/blob/v0.9.0/R/pkg.R#L328
+  if (source %in% c("github", "gitlab", "bitbucket")) {
+    # SCM information is recorded elsewhere
+    repository <- NA_character_
+  } else if (source == "source") {
+    # can't install source packages elsewhere
+    repository <- NA_character_
+    source <- NA_character_
+  } else if (source == "CustomCRANLikeRepository") {
+    # Package was installed from source, but packrat guessed it was installed
+    # from a known repo. We need to check that it's not actually a dev version
+    if (isDevVersion(record, availablePackages)) {
+      repository <- NA_character_
+      source <- NA_character_
+    } else {
+      repository <- findRepoUrl(pkg, availablePackages)
+      source <- findRepoName(repository, repos)
+    }
+  } else if (source %in% c("CRAN", "Bioconductor")) {
+    # shinyapps & posit.cloud will ignore, but connect will use (unless admin
+    # has set up an override)
+    repository <- findRepoUrl(pkg, availablePackages)
+  } else {
+    # Installed from custom repository. Find URL from available.packages()
+    # and then name from repos.
+    repository <- findRepoUrl(pkg, availablePackages)
+    source <- findRepoName(repository, repos)
+  }
+  list(Source = source, Repository = repository)
+}
+
+findRepoName <- function(repository, repos) {
+  idx <- match(repository, repos)
+  names(repos)[idx]
+}
+
+findRepoUrl <- function(pkg, availablePackages) {
+  if (pkg %in% rownames(availablePackages)) {
```

Comment:
Nit: might be slightly safer to check `[, "Package"]` in case there were duplicate row names that got unique-ified?

Reply:
(or can we already guarantee that this wouldn't occur? I'm thinking of things like where `MASS` and recommended packages are available in two parts of the repository, e.g. in `Recommend/4.3.0` or something)

Reply:
I hadn't thought about that problem, thanks!